### PR TITLE
fix: Update TopicOperatorTest to handle async flow correctly

### PR DESCRIPTION
**Pull request by katheris**
### Type of change

_Select the type of your PR_

- Bugfix

### Description

Update the TopicOperatorTest to fix problems with asynchronous tests being completed early before all assertions are run. This fixes issue #7072.

Main changes are:
 - Update the resourceAdded and similar functions so they return a future, rather than creating and flagging a checkpoint
 - Change checkpoints to context.completeNow() where possible
 - Remove the use of CountdownLatch in favour of vertx future composition

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

**Comment by katheris** [comment]
This change is complete, but as you can see three of the tests fail. These are:
 - testReconcile_withResource_withKafka_noPrivate_matching
 - testReconcileMetricsWithPausedTopic
 - testReconcileMetricsDeletedTopic

The first one the error is that the expected metrics for the topic cannot be found. These assertions weren't being run previously due to the checkpoint being flagged before the end of the test.

The second two are both failing within the addResource or deleteResource functions. Previously these functions ran in parallel with the reconcile call so it is unclear what order they are supposed to be in to pass correctly. If I comment out the specific assertions that are failing, the others assertions in those tests do pass.

**Review by SamBarker — COMMENTED** [review]
Certainly an improvement 😁

File: topic-operator/src/test/java/io/strimzi/operator/topic/TopicOperatorTest.java
Comment (katheris) [reply topic-operator/src/test/java/io/strimzi/operator/topic/TopicOperatorTest.java:146]: I think you're correct that we could change this to be a fully synchronous test. I did consider doing that but wondered if it would be better to leave it consistent to reduce future possibilities that people mistakenly copy and paste this test. As I think what happened with the other tests in this class is originally the resourceAdded function was always the last step in the test, but then it got reused in another test where the intent wasn't understood. So that's why I left it as "asynchronous"

File: topic-operator/src/test/java/io/strimzi/operator/topic/TopicOperatorTest.java
Comment (katheris) [reply topic-operator/src/test/java/io/strimzi/operator/topic/TopicOperatorTest.java:165]: I'm not aware of any particular reasons to have them separated vs grouped, if I was writing them from scratch I'd probably choose to group them just so the assertThat statements are easier to pick out as you scan down the code.

File: topic-operator/src/test/java/io/strimzi/operator/topic/TopicOperatorTest.java
Comment (SamBarker) [reply topic-operator/src/test/java/io/strimzi/operator/topic/TopicOperatorTest.java:146]: Seems a reasonable justification to me, let's leave it as is.

**Review by tombentley — APPROVED** [review]
LGTM, thanks @katheris

**Review by scholzj — APPROVED** [review]

**Comment by scholzj** [comment]
@katheris I assume you want to wait for @SamBarker before we merge this?

**Comment by katheris** [comment]
@scholzj yeah, let's wait for him to look over the latest changes

**Pull request merged**
